Add REDIRECT for request errors and established subscriptions (#1534)
moq-wg/moq-transport: draft-ietf-moq-transport.md

  such as language tags, that would aid comprehension by any entity other than
  the one that created the text.

### Redirect Structure {#redirect-structure}

A Redirect provides a way for an endpoint to direct the peer to retry a
request at a different URI and/or for a different Full Track Name.

~~~
Redirect {
  Connect URI Length (vi64),
  Connect URI (..),
  Track Namespace (..),
  Track Name Length (vi64),
  Track Name (..),
}
~~~

* Connect URI: The URI to connect to for this track. If the length is
  zero, the requester SHOULD use the current session's URI. If a server

>   receives a Redirect with a non-zero Connect URI Length it MUST close the

Reviewer: Is this always true?  If two relays contact each other, then one is the client and one is the server, but I can imagine either wanting to do a redirect?

: The PUBLISH or SUBSCRIBE request attempted to create a subscription to a Track
with the same role as an existing subscription.

REDIRECT:
: The request cannot be fulfilled by this endpoint, but could succeed at the
location specified in the Redirect structure. The requester SHOULD establish a
new session to the provided URI (if present) and retry the request using the
Full Track Name from the Redirect (if present). This error code can appear in

> response to SUBSCRIBE, FETCH, TRACK_STATUS, and SUBSCRIBE_NAMESPACE.

Reviewer: Above you mention PUBLISH_NAMESPACE, does it apply to PUBLISH_NAMESPACE?  I assume not?

  after parsing all previous fields. Track Properties are populated in
  TRACK_STATUS_OK; they are empty in REQUEST_UPDATE_OK,
  SUBSCRIBE_NAMESPACE_OK and PUBLISH_NAMESPACE_OK.  If an endpoint
  receives Track Properties in one of these messages it MUST close the
  session with a `PROTOCOL_VIOLATION`.

## REQUEST_ERROR {#message-request-error}

The REQUEST_ERROR message is sent in response to any request (SUBSCRIBE, FETCH,
PUBLISH, SUBSCRIBE_NAMESPACE, PUBLISH_NAMESPACE, TRACK_STATUS, REQUEST_UPDATE).

### Redirect Structure {#redirect-structure}

A Redirect provides a way for an endpoint to direct the peer to retry a
request at a different URI and/or for a different Full Track Name.

~~~
Redirect {
  Connect URI Length (vi64),
  Connect URI (..),
  Track Namespace (..),
  Track Name Length (vi64),
  Track Name (..),
}
~~~

* Connect URI: The URI to connect to for this track. If the length is
  zero, the requester SHOULD use the current session's URI. If a server
  receives a Redirect with a non-zero Connect URI Length it MUST close the
  session with a `PROTOCOL_VIOLATION`.

* Track Namespace: The Track Namespace to use for the redirected request. If

>   both Track Namespace and Track Name have zero length, the redirected request

Reviewer: What if only one is zero length?

## REQUEST_ERROR {#message-request-error}

The REQUEST_ERROR message is sent in response to any request (SUBSCRIBE, FETCH,
PUBLISH, SUBSCRIBE_NAMESPACE, PUBLISH_NAMESPACE, TRACK_STATUS, REQUEST_UPDATE).

### Redirect Structure {#redirect-structure}

A Redirect provides a way for an endpoint to direct the peer to retry a
request at a different URI and/or for a different Full Track Name.

~~~
Redirect {
  Connect URI Length (vi64),
  Connect URI (..),

>   Track Namespace (..),
>   Track Name Length (vi64),
>   Track Name (..),

Reviewer 2: ```suggestion
  [Track Namespace (..),
  Track Name Length (vi64),
  Track Name (..)],
```

Author: I left them required, but said that 0 length Track Namespace / Name means "use the same one in the original request"

}
~~~

* Connect URI: The URI to connect to for this track. If the length is
  zero, the requester SHOULD use the current session's URI. If a server
  receives a Redirect with a non-zero Connect URI Length it MUST close the
  session with a `PROTOCOL_VIOLATION`.

* Track Namespace: The Track Namespace to use for the redirected request.

* Track Name: The Track Name to use for the redirected request. Track Name
  is not meaningful for namespace-scoped requests (SUBSCRIBE_NAMESPACE,
  PUBLISH_NAMESPACE) and MUST be empty; an endpoint that receives a non-empty
  Track Name in a Redirect for a namespace-scoped request MUST close the session

>   with a `PROTOCOL_VIOLATION`.

Reviewer 2: I am not sure how a relay would know the right FullTrackName which is application scoped.  I

Author: @suhasHere this is done via configuration rules typically, it's not in-band

Reviewer 2: Then we should make this an optional field ?

Author: Done, see above

EXCESSIVE_LOAD (0x9):
: The publisher is overloaded and is terminating the subscription.

## REDIRECT {#message-redirect}

An endpoint sends a REDIRECT message on a bidirectional request stream to

> indicate that the peer should re-establish the request at a different location.

Reviewer: ```suggestion
indicate that the peer can re-establish the request on a different session.
```
